### PR TITLE
MGMT-12301: refactor the progress bar to use state machine

### PR DESCRIPTION
**Pull request by slaviered**
WIP

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/host/statemachine.go [hunk 1]
```diff
@@ -42,6 +43,86 @@ func NewHostStateMachine(sm stateswitch.StateMachine, th TransitionHandler) stat
...
+		PostTransition:   th.PostHostProgress,
+		Documentation: stateswitch.TransitionRuleDoc{
+			Name:        "host progress changed",
+			Description: "Change day2 host state to HostStatusAddedToExistingCluster when it reached stage Done. (i.e. the end of SAAS flow for day2 installation)",
+		},
+	})
+
+	sm.AddTransitionRule(stateswitch.TransitionRule{
+		TransitionType: TransitionTypeHostProgress,
+		Condition:      th.IsHostInDone,
```
[the remark this answers is not in the document]
Comment (slaviered) [reply internal/host/statemachine.go:78]: The order matters. I also added a comment above the default rule to keep it last in the chain

File: internal/host/statemachine.go [hunk 2]
```diff
@@ -42,6 +43,86 @@ func NewHostStateMachine(sm stateswitch.StateMachine, th TransitionHandler) stat
...
+		SourceStates: []stateswitch.State{
+			stateswitch.State(models.HostStatusInstallingInProgress),
+			stateswitch.State(models.HostStatusInstalling),
+			stateswitch.State(models.HostStatusInstallingPendingUserAction),
+		},
+		DestinationState: stateswitch.State(models.HostStatusInstallingInProgress),
+		PostTransition:   th.PostHostProgress,
+		Documentation: stateswitch.TransitionRuleDoc{
+			Name:        "host progress changed",
+			Description: "Change host state to installed when it reached stage Done",
```
[the remark this answers is not in the document]
Comment (slaviered) [reply internal/host/statemachine.go:104]: Fixed all descriptions

**Review by avishayt — APPROVED** [review]

**Pull request merged**